homeassistant/components/freebox/router.py

```diff
@@ -0,0 +1,247 @@
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry):
+        """Initialize a Freebox router."""
+        self.hass = hass
+        self._entry = entry
+        self._host = entry.data[CONF_HOST]
+        self._port = entry.data[CONF_PORT]
+
+        self._api: Freepybox = None
+        self._name = None
+        self._mac = None
+        self._sw_v = None
+
+        # Devices
+        self._devices: Dict[str, FreeboxDevice] = {}
+
+        # Sensors
+        self._temperature_sensors: Dict[str, FreeboxSensor] = {}
+        self._connection_sensors: Dict[str, FreeboxSensor] = {}
+        self._attrs = {}
+
+    async def setup(self) -> None:
+        """Set up a Freebox router."""
+        freebox_dir = self.hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+
+        if not os.path.exists(freebox_dir.path):
+            await self.hass.async_add_executor_job(os.makedirs, freebox_dir.path)
+
+        token_file = self.hass.config.path(
+            f"{freebox_dir.path}/{slugify(self._host)}.conf"
+        )
+
+        self._api = Freepybox(APP_DESC, token_file, API_VERSION)
+
+        await self._api.open(self._host, self._port)
+
+        # System
+        fbx_config = await self._api.system.get_config()
+        self._mac = fbx_config["mac"]
+        self._name = fbx_config["model_info"]["pretty_name"]
+        self._sw_v = fbx_config["firmware_version"]
+
+        # Devices & sensors
+        await self.update_all()
+        async_track_time_interval(self.hass, self.update_all, SCAN_INTERVAL)
+
+    async def update_all(self, now=None) -> None:
+        """Update all Freebox platforms."""
+        await self.update_sensors()
+        await self.update_devices()
+
+    async def update_devices(self) -> None:
+        """Update Freebox devices."""
+        if self._api is None:
+            return
+
+        fbx_devices: Dict[str, any] = await self._api.lan.get_hosts_list()
+
+        # Adds the Freebox itself
+        fbx_devices.append(
+            {
+                "primary_name": self.name,
+                "l2ident": {"id": self.mac},
+                "vendor_name": self.manufacturer,
+                "host_type": "router",
+                "active": True,
+                "attrs": self._attrs,
+            }
+        )
+
+        for fbx_device in fbx_devices:
+            device_name = fbx_device["primary_name"].strip() or DEFAULT_DEVICE_NAME
+            device_mac = fbx_device["l2ident"]["id"]
+
+            if self._devices.get(device_mac) is not None:
+                # Seen device -> updating
+                _LOGGER.debug("Updating Freebox device: %s", device_name)
+                self._devices[device_mac].update_state(fbx_device)
+            else:
+                # New device, should be unique
+                _LOGGER.debug(
+                    "Adding Freebox device: %s [MAC: %s]", device_name, device_mac,
+                )
+                self._devices[device_mac] = FreeboxDevice(fbx_device)
```

Comment:
Creating entities should happen by control of their platforms. Please move the creation to the platform.

It's to to signal the platform from here to create and add an entity.

Reply:
See eg unifi device tracker.

Reply:
I've seen it because needed for iCloud #31909 (devices not reachable at first API fetch), I think the code is awful.

Will it also permits to add `FreeboxDevice`s after the component setup ?

Reply:
That's the whole point of the signal. Otherwise we could just create all entities from some existing initial data in the platform.

Reply:
It looks some time bug I get it !

Very nice addition !